File: codemem/coordinator_invites.py
from __future__ import annotations

import base64
import json
from typing import TypedDict
from urllib.parse import parse_qs, quote, urlparse


class InvitePayload(TypedDict):
    v: int
    kind: str
    coordinator_url: str
    group_id: str
    policy: str
    token: str
    expires_at: str
    team_name: str | None


def encode_invite_payload(payload: InvitePayload) -> str:
    raw = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def decode_invite_payload(value: str) -> InvitePayload:
    padding = "=" * (-len(value) % 4)
    data = json.loads(base64.urlsafe_b64decode(value + padding).decode("utf-8"))
    if not isinstance(data, dict):
        raise ValueError("invalid invite payload")
    return data  # type: ignore[return-value]


def invite_link(encoded_payload: str) -> str:
    return f"codemem://join?invite={quote(encoded_payload)}"


def extract_invite_payload(value: str) -> str:
    raw = value.strip()
    if raw.startswith("codemem://"):
        parsed = urlparse(raw)
        invite = parse_qs(parsed.query).get("invite", [""])[0]
        if not invite:
            raise ValueError("invite payload missing from link")
        return invite
    return raw

File: packages/core/src/coordinator-invites.ts
/**
 * Invite link encoding/decoding for the coordinator join flow.
 *
 * Ported from codemem/coordinator_invites.py.
 */

export interface InvitePayload {
	v: number;
	kind: string;
	coordinator_url: string;
	group_id: string;
	policy: string;
	token: string;
	expires_at: string;
	team_name: string | null;
}

/** Encode an invite payload as a compact base64url string (no padding). */
export function encodeInvitePayload(payload: InvitePayload): string {
	const json = JSON.stringify(payload);
	const bytes = new TextEncoder().encode(json);
	const base64 = Buffer.from(bytes).toString("base64url");
	// base64url from Buffer already omits padding — strip defensively
	return base64.replace(/=+$/, "");
}

/** Decode a base64url invite payload string back to an InvitePayload. */
export function decodeInvitePayload(value: string): InvitePayload {
	// Re-add padding that was stripped
	const padded = value + "=".repeat((4 - (value.length % 4)) % 4);
	const json = Buffer.from(padded, "base64url").toString("utf-8");
	const data: unknown = JSON.parse(json);
	if (typeof data !== "object" || data === null) {
		throw new Error("invalid invite payload");
	}
	return data as InvitePayload;
}

/** Build a `codemem://join?invite=...` link from an encoded payload. */
export function inviteLink(encodedPayload: string): string {
	return `codemem://join?invite=${encodeURIComponent(encodedPayload)}`;
}

/**
 * Extract the raw encoded payload from either a `codemem://` link or a raw value.
 *
 * Throws if a `codemem://` link is provided but has no `invite` query parameter.
 */
export function extractInvitePayload(value: string): string {
	const raw = value.trim();
	if (raw.startsWith("codemem://")) {
		const url = new URL(raw);
		const invite = url.searchParams.get("invite") ?? "";
		if (!invite) {
			throw new Error("invite payload missing from link");
		}
		return invite;
	}
	return raw;
}
